tests/components/ecovacs/conftest.py

```diff
@@ -55,10 +84,46 @@ def mock_authenticator_authenticate(mock_authenticator: Mock) -> AsyncMock:
 
 
 @pytest.fixture
-def mock_api_client(mock_authenticator: Mock) -> Mock:
-    """Mock the API client."""
+def mock_mqtt_client(mock_authenticator: Mock) -> Mock:
+    """Mock the MQTT client."""
     with patch(
-        "homeassistant.components.ecovacs.controller.ApiClient",
-        return_value=Mock(spec_set=ApiClient),
-    ) as mock_api_client:
-        yield mock_api_client.return_value
+        "homeassistant.components.ecovacs.controller.MqttClient",
+        autospec=True,
+    ) as mock_mqtt_client:
+        client = mock_mqtt_client.return_value
+        client._authenticator = mock_authenticator
+        client.subscribe.return_value = lambda: None
+        yield client
+
+
+@pytest.fixture
+def mock_device_execute() -> AsyncMock:
+    """Mock the device execute function."""
+    with patch.object(
+        Device, "_execute_command", return_value=True
+    ) as mock_device_execute:
+        yield mock_device_execute
+
+
+@pytest.fixture
+async def init_integration(
+    hass: HomeAssistant,
+    mock_config_entry: MockConfigEntry,
+    mock_authenticator: Mock,
+    mock_mqtt_client: Mock,
+    mock_device_execute: AsyncMock,
+) -> MockConfigEntry:
+    """Set up the Ecovacs integration for testing."""
+    mock_config_entry.add_to_hass(hass)
+
+    await hass.config_entries.async_setup(mock_config_entry.entry_id)
+    await hass.async_block_till_done()
+    return mock_config_entry
+
+
+@pytest.fixture
+def controller(
+    hass: HomeAssistant, init_integration: MockConfigEntry
+) -> EcovacsController:
+    """Get the controller for the config entry."""
+    return hass.data[DOMAIN][init_integration.entry_id]
```

Comment:
This is an integration detail that we should not use in the tests. Try to patch the library instead to get access to the relevant resource.